### ISSUE TITLE
build command does not copy the actual content of css files

### ISSUE THREAD

**Issue by rokde** [jekyll/jekyll#4027]
Jekyll build command copies an old content version of a static css file within my source folder. Does it have any file cache or so? Hash sum of css file changes so I do not know where the hell jekyll does copy the old version content of the css file.

I use jekyll within a grunt-jekyll task, but the direct call/use of jekyll does not change this behaviour.

Please fix that problem. Or give me help for debugging that thing.

**Comment by envygeeks** [comment]
What is grunt doing?

**Comment by rokde** [comment]
It handles markdown files and creates liquid style templates...afterwards jekyll gets a configuration and renders liquid files to html (everything fine) but the file copies result in false content. It takes the old content which is not in the source anymore right now.
The environment is the scotch box (https://box.scotch.io/), maybe this is the problem?

**Comment by envygeeks** [comment]
@rokde can you provide a working example of your setup so I can test this? I'm a full time Ubuntu user so if you can reproduce it I will see if I can to track down where it's happening.

**Comment by bexelbie** [comment]
I am seeing something similar in boot2docker using the stock jekyll container.  My css file was updated.  The file that lands in _site is the old file with some padding to fill out the bytes of my new additions.  Changes and deletions were not made, just additions.

**Comment by envygeeks** [comment]
I do not wish to get on my Mac.  @jekyll/core one of you want to play with this?

**Comment by parkr** [comment]
@bexelbie Would you run that command with `--verbose` and paste the output for me in a gist and link it here?

**Comment by bexelbie** [comment]
@parkr https://gist.github.com/bexelbie/b940ca4e54c39143836f

**Comment by parkr** [comment]
@bexelbie Interesting. Have you been able to reproduce this on Linux outside of the docker container? It's always possible that the updated files aren't being copied.

**Comment by envygeeks** [comment]
I see quite a few filesystem bugs for boot2docker, what are you using as the middle-man? VirtualBox, VMWare, VMWare Fusion, Parallels (ew?)

**Comment by bexelbie** [comment]
@parkr I haven't tried it outside of the container.  Not to be obstructionist, but I am using the container deliberately to avoid a local install.

@envygeeks I am using Virtualbox on the mac.  I am not able to recreate it on a Linux box using local docker.  I will keep a pin in this to try it via VirtualBox and libvirt on the Linux box, however that won't happen today.

Can we pin this directly to one of the reported bugs.  I find it surprising this is an edge case, so I would have expected this to hit more people.

**Comment by envygeeks** [comment]
boot2docker doesn't work on Linux, it's for Windows and OS X only.  Unless you mean `docker-machine`.  Which leads to my next suggestion, try using docker-machine which (and I might be wrong) is what they are pushing to anyways because it gets much more love than boot2docker.  As a point last I remember the Mac installer doesn't even do boot2docker anymore, it's straight up docker-machine -> docker, docker-compose -> docker-machine -> docker.

**Comment by envygeeks** [comment]
Apparently they also provide a migration path: https://docs.docker.com/engine/installation/mac/#migrate-from-boot2docker

**Comment by bexelbie** [comment]
@envygeeks the most recent tests were with docker-machine.  Both boot2docker and docker-machine use the same base iso and the same virtualization provider.

**Comment by envygeeks** [comment]
I understand that.  I'm not new to Docker, it was me who built the images you use :wink: however, they approach the problem two different ways.  That said, it works on Linux, our build server uses Docker-Machine and it would have failed if it couldn't build and detect changes.

**Comment by envygeeks** [comment]
I honestly don't know what could be going on.  Try booting it to the background and logging in to see if it makes the changes on the machine it self:

```
docker run --rm --label=jekyll --volume=$(pwd):/srv/jekyll -dit -p 4000:4000 jekyll/jekyll sh
docker run -it jekyll b
docker exec -it cat path/to/file.css
```

**Comment by envygeeks** [comment]
> I honestly don't know what could be going on.

Let me rephrase that, I honestly don't know what could be going on without logs.  I can come up with a thousand reasons as to what would be going wrong but that would be chasing the golden goose and waste all people's time.

**Comment by bexelbie** [comment]
@envygeeks I wasn't trying to question your, or anyone's, skills with docker or vms.  I was just pointing out that even with docker-machine the operative bits in the background seem to be the same.  I will look at the filesystem on the vm.  I am not sure why that didn't occur to me before, but it is a great idea.

What logs would you like to see in addition to the verbose I pasted above via gist?

**Comment by envygeeks** [comment]
VirtualBox logging with `+shared_folders.e.f` should give some hints (it's been a while since I've had to do that so I don't quite remember.)

**Comment by envygeeks** [comment]
Yeah, it looks like this might work to give you some hints so you can sort through and see where to go:
there is a log in `$HOME/.docker/machine/machines/machinename/machinename/Logs/VBox.log` according to Docker-Machine, they always create one, which means you should be able to do:

```
VBOX_LOG_FLAGS=thread
VBOX_LOG=+shared_folders.e.f
```

To get some information.

**Comment by envygeeks** [comment]
You can also run `docker-machine` with `--debug` and get even more information.

**Comment by bexelbie** [comment]
So I did a bit of "inception testing".  I am able to see the corrupted file(s) at every level:
* on the host
* in the VM
* in the docker container running in the VM

I tried exporting those two ENV vars but got no additional output or relevant output in the VBOX.log.  Any hints?

**Comment by bexelbie** [comment]
A bit more potential information.  This appears to be a manifestation of this bug in VirtualBox https://www.virtualbox.org/ticket/9069.

My working theory is that somehow the copy done by regenerate is triggering the file caching bug.  

I don't know that this should be fixed in jekyll.  However, I'd love to be able to test this theory.  I just don't know my way around the jekyll code base.  I'd like to have jekyll build treat all files as "built" meaning don't copy anything, actually open a new file handle and write it out.  If that worked, it would confirm this is that bug.

**Comment by ChristophRob** [comment]
Hey guys, 
we have the same Problem with our docker. A new css/js file copies perfectly into our static folder but after we change it, the changes wont get updated. The old content of the css / js  files gets copied. It seems to have something to do with the number of characters. So if i erase the 13000 line css file and put a short random text in It, only the first few lines of the old css file are getting copied. 

No Problem with html files though. 

The problem only occurs on OS X with docker. Tried Docker Toolbox and the older boot2docker with VirtualBox. Linux works fine. Interestingly we are using not Jekyll, but Django!

Hope you guys have any ideas how to fix it

Cheers
Christoph

**Comment by ChristophRob** [comment]
we fixed it in django_uwsgi_dev.ini with disable-sendfile = true

**Comment by jakubpawlowicz** [comment]
Same here, VirtualBox 5.0.14, docker-machine 0.6, Docker 1.10 on OS X 10.11 (host) running Debian Jessie (guest). It worked fine on OS X 10.10 and VirtualBox 5.0.4 but annoyingly it does not anymore.

**Comment by jakubpawlowicz** [comment]
Managed to work around this issue by using https://github.com/nlf/dlite instead of VirtualBox, however it's Mac only.

**Comment by justinschier** [comment]
I've been having a very similar issue using the latest Jekyll 3.1.6.

I'm using a docker machine (with Kitematic), and I was having the strangest issue with a Javascript file. A chopped version of the file kept getting put into _site/scripts. If I would edit the source file, a differently chopped version would get copied. And even if I removed some text from the top of the file, it was still there in the copy. Super weird. After bouncing my entire machine, the behavior starts up again. No clue why it's affecting only some files.

I am using `jekyll serve`, with the source code living on my local Mac drive and mapped from the docker machine. After making a file change, I have to manually do a `touch scripts/myfile.js` to get jekyll to regen. 

I see now @bexelbie has pointed to a caching bug in VirtualBox, which sounds probable. I just don't understand any of the workarounds or how to get past this.

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least
one month.

The resources of the Jekyll team are limited, and so we are asking for your help.

If you can still reproduce this error on the <code>3.1-stable</code> or <code>master</code> branch,
please reply with all of the information you have about it in order to keep the issue open.

If this is a feature request, please consider building it first as a plugin. Jekyll 3 introduced
[hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout
the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be
built as a plugin, then please provide more information about why in order to keep this issue open.

Thank you for all your contributions.

**Comment by Starefossen** [comment]
I can confirm that this is still a problem with `3.1-stable`.

**Comment by parkr** [comment]
@Starefossen Would you please provide more information? Or submit a pull request with a failing test so we may fix it. There is some discussion above about this being a VirtualBox bug, but we cannot move forward without some more information.

**Comment by justinschier** [comment]
I have switched to the Docker for Mac beta (away from Kitematic) and the issue is no longer happening to me. I believe it was related to some kind of bug in file system caching when using VirtualBox and Kitematic.

Unfortunately the jekyll build now takes 2X as long with Docker for Mac beta 1.12.0-rc2-beta16 (build: 9493).

**Comment by envygeeks** [comment]
@justinschier see https://github.com/jekyll/docker/issues/103 for a discussion on that.

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least two months.

The resources of the Jekyll team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>3.1-stable</code> or <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, please consider building it first as a plugin. Jekyll 3 introduced [hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be built as a plugin, then please provide more information about why in order to keep this issue open.

This issue will automatically be closed in two months if no further activity occurs. Thank you for all your contributions.

**Comment by WangKunKun** [comment]
jekyll 3.1.6 still can't modify main.css .

new main.css --> jekyll build --> old main.css

**Comment by jekyllbot** [comment]
Hello! We are utilizing a new workflow in our issues and pull requests. Affinity teams have been setup to allow community members to hear about pull requests that may be interesting to them. When a new issue or pull request comes in, we are asking that the author mention the appropriate affinity team. I then assign a random "team captain" or two to the issue who is in charge of triaging it until it is closed or passing it off to another captain. In order to move forward with this new workflow, we need to know: which of the following teams best fits your issue or contribution?

- `@jekyll/build` – Concerned with the Jekyll build pipeline, from reading files, rendering them, and writing them. Includes incremental regeneration.
- `@jekyll/documentation` – Concerned with making jekyllrb.com and other Jekyll documentation great.
- `@jekyll/ecosystem` – Concerned with the fostering of a vibrant ecosystem. Plugins, themes, etc.
- `@jekyll/performance` – Concerned with Jekyll's speed.
- `@jekyll/stability` – Concerned with all things stability, including configuration, defaults, deprecation, etc.
- `@jekyll/windows` – Concerned with Windows compatibility. Can provide testing and guidance for Windows-related bug fixes and enhancements.

Mention one of these teams in a comment below and we'll get this sorted. Thanks!

**Comment by parkr** [comment]
> new main.css --> jekyll build --> old main.css

@WangKunKun Which file, exactly are you editing? You're not editing anything inside `_site`, right?

**Comment by jekyllbot** [comment]
Hello! We are utilizing a new workflow in our issues and pull requests. Affinity teams have been setup to allow community members to hear about pull requests that may be interesting to them. When a new issue or pull request comes in, we are asking that the author mention the appropriate affinity team. I then assign a random "team captain" or two to the issue who is in charge of triaging it until it is closed or passing it off to another captain. In order to move forward with this new workflow, we need to know: which of the following teams best fits your issue or contribution?

- `@jekyll/build` – Concerned with the Jekyll build pipeline, from reading files, rendering them, and writing them. Includes incremental regeneration.
- `@jekyll/documentation` – Concerned with making jekyllrb.com and other Jekyll documentation great.
- `@jekyll/ecosystem` – Concerned with the fostering of a vibrant ecosystem. Plugins, themes, etc.
- `@jekyll/performance` – Concerned with Jekyll's speed.
- `@jekyll/stability` – Concerned with all things stability, including configuration, defaults, deprecation, etc.
- `@jekyll/windows` – Concerned with Windows compatibility. Can provide testing and guidance for Windows-related bug fixes and enhancements.

Mention one of these teams in a comment below and we'll get this sorted. Thanks!

**Comment by mads-thines** [comment]
I have the same problem.
1. Delete the old _site folder to make sure it's removed.
2. Compiling my Sass files into a bundle.css (Which includes new changes)
3. jekyll serve 
4. New _site folder is being generated and it have the old css.

I'm not using Docker.

**Comment by fraank** [comment]
Same here. I'm using Vagrant on OSX.

**Comment by envygeeks** [comment]
Whether you are using Docker or not, the bug does not originate in Docker, the bug originates in VirtualBox.

**Comment by mads-thines** [comment]
I think I found out what was causing the problem for me.
The repository I was working had the following command in the README file:   
`$ bundle exec jekyll serve -w`

And I think that might have caused the problem for me.
I switched to the following command and the problem went away.

`$ jekyll serve -w`

Not sure if it's related, but it solved my problem.

**Comment by mads-thines** [comment]
I found out the folder that including the bundle.css wasn't allowed to start with the "_" as the other folders. I renamed my _dist > dist and it worked.

I'm not sure if it's really related, but maybe it can help somebody.

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least two months.

The resources of the Jekyll team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>3.1-stable</code> or <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, please consider building it first as a plugin. Jekyll 3 introduced [hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be built as a plugin, then please provide more information about why in order to keep this issue open.

This issue will automatically be closed in two months if no further activity occurs. Thank you for all your contributions.

**Comment by lonectzn** [comment]
VirtualBox has had outstanding tickets that I think are related to this for a while. See [#819](https://www.virtualbox.org/ticket/819) and [#9069](https://www.virtualbox.org/ticket/9069).

I'm not sure how jekyll is transferring assets into the _site folder, but if it is using `sendfile` or `mmap` then it's possible it is affected by this defect. The permanent solution is within VirtualBox's vboxfs, but considering those tickets have been up for about 8 years, good luck waiting on that. 

An easier workaround, and one used in Apache, is just to provide an option not to use sendfile.

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least two months.

The resources of the Jekyll team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>3.3-stable</code> or <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, please consider building it first as a plugin. Jekyll 3 introduced [hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be built as a plugin, then please provide more information about why in order to keep this issue open.

This issue will automatically be closed in two months if no further activity occurs. Thank you for all your contributions.
